whatthelog/prefixtree/graph.py

```diff
@@ -96,6 +96,33 @@ def merge_states(self, state1: State, state2: State) -> None:
         del self.state_indices_by_id[id(state2)]
         del state2
 
+    def full_merge_states(self, s1: State, s2: State):
+        """
+        Fully merges two states and removes non-determinism.
+        :param s1: One of the two states to merge.
+        :param s2: One of the two states to merge.
+        """
+        if s2 is None or s1 is None:
+            return
+
+        # Trivially merge the states
+        self.merge_states(s1, s2)
+
+        # Remove non-determinism in the merged state's children by merging them.
+        current, changed = self.merge_equivalent_children(s1)
+
+        # Get the current state's children
+        parents = self.get_incoming_states(current)
+
+        # For each parent
+        while len(parents) > 0:
+            # Remove non-determinism in the parent
+            current, changed = self.merge_equivalent_children(parents.pop())
```

Comment:
Yes it will. A node with a self loop is now also its own parent and child, and there will thus be overlap in the node's children in the case that you describe. I will push an additional test with this particular case.